### PR TITLE
feat: add rename_app interface 

### PR DESCRIPTION
**Pull request by GehaFearless**
issue: https://github.com/apache/incubator-pegasus/issues/1185

I want add a couple rpc to solve rename app.

### CODE REVIEW

File: src/meta/server_state.cpp [hunk 1]
```diff
@@ -1259,6 +1259,90 @@ void server_state::drop_app(dsn::message_ex *msg)
     }
 }
 
+void server_state::do_app_rename(configuration_rename_app_rpc rpc)
+{
+    int32_t app_id = rpc.request().app_id;
+    const auto &new_app_name = rpc.request().new_app_name;
+
+    auto target_app = get_app(app_id);
```
Comment (acelyc111) [inline src/meta/server_state.cpp:1267]: deal when app not found

File: src/meta/server_state.cpp [hunk 2]
```diff
@@ -1259,6 +1259,90 @@ void server_state::drop_app(dsn::message_ex *msg)
...
+               new_app_name);
+
+    auto ainfo = *(reinterpret_cast<app_info *>(target_app.get()));
+    ainfo.app_name = new_app_name;
+    auto app_path = get_app_path(*target_app);
+    do_update_app_info(app_path, ainfo, [this, target_app, rpc](error_code ec) mutable {
+        const auto &new_app_name = rpc.request().new_app_name;
+        zauto_write_lock l(_lock);
+
+        CHECK(ec == ERR_OK,
```
Comment (acelyc111) [inline src/meta/server_state.cpp:1280]: CHECK_EQ

File: src/meta/server_state.cpp [hunk 3]
```diff
@@ -1259,6 +1259,90 @@ void server_state::drop_app(dsn::message_ex *msg)
...
+                target_app->status == app_status::AS_RECALLING) {
+                response.err = ERR_BUSY_CREATING;
+            } else if (target_app->status == app_status::AS_DROPPING) {
+                response.err = ERR_BUSY_DROPPING;
+            } else if (target_app->status == app_status::AS_DROPPED) {
+                response.err = ERR_APP_DROPPED;
+            } else {
+                // AS_INVALID/AS_CREATE_FAILED/AS_DROP_FAILED
+                response.err = ERR_INVALID_STATE;
+            }
```
Comment (acelyc111) [inline src/meta/server_state.cpp:1331]: how about use switch..case.. ?

File: src/meta/server_state.cpp [hunk 4]
```diff
@@ -1259,6 +1259,90 @@ void server_state::drop_app(dsn::message_ex *msg)
...
+        } else {
+            if (_exist_apps.find(new_app_name) != _exist_apps.end()) {
+                response.err = ERR_INVALID_PARAMETERS;
+            } else {
+                do_rename = true;
```
Comment (acelyc111) [inline src/meta/server_state.cpp:1336]: reduce if-else statements by:
```
do {
  if cond1
    ...
    break;
  if cond2
    ...
    break;
  ...
} while (false);
```

File: src/replica/replica_config.cpp [hunk 5]
```diff
@@ -1070,6 +1071,25 @@ void replica::on_config_sync(const app_info &info,
     }
 }
 
+void replica::update_app_name(const std::string app_name)
```
Comment (acelyc111) [inline src/replica/replica_config.cpp:1074]:
```suggestion
void replica::update_app_name(const std::string &app_name)
```

File: src/replica/replica_config.cpp [hunk 6]
```diff
@@ -1070,6 +1071,25 @@ void replica::on_config_sync(const app_info &info,
...
+{
+    if (app_name == _app_info.app_name) {
+        return;
+    }
+
+    auto old_app_name = _app_info.app_name;
+    _app_info.app_name = app_name;
+
+    auto ec = store_app_info(_app_info);
+    CHECK_EQ_PREFIX_MSG(ec,
```
Comment (acelyc111) [inline src/replica/replica_config.cpp:1084]:
```suggestion
    CHECK_EQ_PREFIX_MSG(store_app_info(_app_info),
```

File: src/replica/replica.h [hunk 7]
```diff
@@ -496,6 +496,7 @@ class replica : public serverlet<replica>, public ref_counter, public replica_ba
     clear_on_failure(replica_stub *stub, replica *rep, const std::string &path, const gpid &pid);
 
     void update_app_max_replica_count(int32_t max_replica_count);
+    void update_app_name(const std::string app_name);
```
Comment (acelyc111) [inline src/replica/replica.h:499]:
```suggestion
    void update_app_name(const std::string &app_name);
```

File: src/client/replication_ddl_client.cpp [hunk 8]
```diff
@@ -1738,5 +1738,27 @@ replication_ddl_client::set_max_replica_count(const std::string &app_name,
...
 }
 
+error_with<configuration_rename_app_response>
+replication_ddl_client::rename_app(const std::string &old_app_name, const std::string &new_app_name)
+{
+    if (old_app_name.empty() ||
+        !std::all_of(old_app_name.cbegin(),
+                     old_app_name.cend(),
+                     (bool (*)(int))replication_ddl_client::valid_app_char))
+        return FMT_ERR(ERR_INVALID_PARAMETERS, "invalid old_app_name.");
```
Comment (acelyc111) [inline src/client/replication_ddl_client.cpp:1748]: It seems this code is duplicated some times, could you add a macros like:
```
VALIDATE_TABLE_NAME(table_name, msg)  \
  if (table_name.empty() || ...) {    \
    return FMT_ERR(ERR_INVALID_PARAMETERS, #msg);
```

you can do it in another patch and refactor some old code.

File: src/client/replication_ddl_client.cpp [hunk 9]
```diff
@@ -1738,5 +1738,27 @@ replication_ddl_client::set_max_replica_count(const std::string &app_name,
...
+    if (!std::all_of(new_app_name.cbegin(),
+                     new_app_name.cend(),
+                     (bool (*)(int))replication_ddl_client::valid_app_char)) {
+        return FMT_ERR(ERR_INVALID_PARAMETERS,
+                       "Not support format for new_app_name. You can use "
+                       "letters and numbers, and also char in ['.', ':', "
+                       "'_'].");
```
Comment (acelyc111) [inline src/client/replication_ddl_client.cpp:1755]: How about simplified by "Only 0-9a-zA-Z.:_ are valid" ? You can append it to the message in the macro mentioned above.

File: src/meta/server_state.cpp [hunk 10]
```diff
@@ -1259,6 +1259,111 @@ void server_state::drop_app(dsn::message_ex *msg)
     }
 }
 
+void server_state::do_app_rename(configuration_rename_app_rpc rpc)
```
Comment (acelyc111) [inline src/meta/server_state.cpp:1262]: How about rename to `rename_app_unlock`? the `_unlock` postfix is often used for the functions which must hold a lock by caller.

File: src/meta/server_state.cpp [hunk 11]
```diff
@@ -1259,6 +1259,111 @@ void server_state::drop_app(dsn::message_ex *msg)
...
+    auto app_path = get_app_path(*target_app);
+
+    do_update_app_info(app_path, ainfo, [this, target_app, rpc](error_code ec) mutable {
+        const auto &new_app_name = rpc.request().new_app_name;
+        zauto_write_lock l(_lock);
+
+        CHECK_EQ(ec, ERR_OK);
+
+        const auto old_app_name = target_app->app_name;
+        target_app->app_name = new_app_name;
```
Comment (empiredan) [inline src/meta/server_state.cpp:1295]: Should we check again whether the new table name has been used, since this is called asynchronously, maybe a new table named `new_app_name` has been created successfully ?

File: src/meta/server_state.cpp [hunk 12]
```diff
@@ -1259,6 +1259,111 @@ void server_state::drop_app(dsn::message_ex *msg)
...
 
+void server_state::do_app_rename(configuration_rename_app_rpc rpc)
+{
+    const auto &old_app_name = rpc.request().old_app_name;
+    const auto &new_app_name = rpc.request().new_app_name;
+
+    zauto_read_lock l;
+    auto target_app = get_app(old_app_name);
+
+    if (target_app == nullptr) {
```
Comment (acelyc111) [inline src/meta/server_state.cpp:1270]: target_app has been checked in `rename_app` under a lock, how can this situaction happened?

File: src/meta/test/meta_app_operation_test.cpp [hunk 13]
```diff
@@ -795,5 +809,28 @@ TEST_F(meta_app_operation_test, recover_from_max_replica_count_env)
...
+
+    const std::string app_name_3 = APP_NAME + "_rename_3";
+
+    auto resp = rename_app(app_name_1, app_name_2);
+    ASSERT_EQ(resp.err, ERR_INVALID_PARAMETERS);
+
+    resp = rename_app(app_name_1, app_name_3);
+    ASSERT_EQ(resp.err, ERR_OK);
+    app = find_app(app_name_3);
+    ASSERT_EQ(app->app_id, app_id_1);
```
[the remark this answers is not in the document]
Comment (GehaFearless) [reply src/meta/test/meta_app_operation_test.cpp:832]: thx

File: src/meta/server_state.cpp [hunk 14]
```diff
@@ -1259,6 +1259,111 @@ void server_state::drop_app(dsn::message_ex *msg)
...
+
+    switch (target_app->status) {
+    case app_status::AS_AVAILABLE: {
+        if (_exist_apps.find(new_app_name) != _exist_apps.end()) {
+            response.err = ERR_INVALID_PARAMETERS;
+            oss << "ERROR: app(" << new_app_name << ") already exist! check it!" << std::endl;
+            response.hint_message += oss.str();
+            return;
+        }
+        do_rename = true;
```
[the remark this answers is not in the document]
Comment (GehaFearless) [reply src/meta/server_state.cpp:1342]: For a precise function，also keep style with others and be good for transformation and optimization.

File: src/meta/test/meta_app_operation_test.cpp [hunk 15]
```diff
@@ -795,5 +809,37 @@ TEST_F(meta_app_operation_test, recover_from_max_replica_count_env)
...
+    app = find_app(app_name_2);
+    CHECK(app, "app({}) does not exist", app_name_2);
+    auto app_id_2 = app->app_id;
+
+    const std::string app_name_3 = APP_NAME + "_rename_3";
+
+    // case 1: new_app_name table exist
+    auto resp = rename_app(app_name_1, app_name_2);
+    ASSERT_EQ(ERR_INVALID_PARAMETERS, resp.err);
+
```
Comment (empiredan) [inline src/meta/test/meta_app_operation_test.cpp:831]: Could add a case that if the new app name has invalid character.

File: src/meta/server_state.cpp [hunk 16]
```diff
@@ -1259,6 +1259,105 @@ void server_state::drop_app(dsn::message_ex *msg)
...
+
+    auto target_app = get_app(old_app_name);
+    auto app_id = target_app->app_id;
+
+    auto ainfo = *(reinterpret_cast<app_info *>(target_app.get()));
+    ainfo.app_name = new_app_name;
+    auto app_path = get_app_path(*target_app);
+
+    target_app->app_name = new_app_name;
+    _exist_apps.emplace(new_app_name, target_app);
```
Comment (empiredan) [inline src/meta/server_state.cpp:1277]: Why not add new app name after updating remote storage ? That will be an atomic operation including both adding new table name and removing old table name.

Comment (GehaFearless) [reply src/meta/server_state.cpp:1277]: Maybe a new table named new_app_name has been created successfully，So add it firstly.

File: src/meta/test/meta_app_operation_test.cpp (hunk 15 above)
Comment (GehaFearless) [reply src/meta/test/meta_app_operation_test.cpp:831]: Valid character checking in `ddl_client`, server not check it. So I think could ignore it here.

File: src/meta/test/meta_app_operation_test.cpp [hunk 17]
```diff
@@ -795,5 +809,28 @@ TEST_F(meta_app_operation_test, recover_from_max_replica_count_env)
     verify_app_max_replica_count(APP_NAME, new_max_replica_count);
 }
 
+TEST_F(meta_app_operation_test, rename_app)
+{
+    const std::string app_name_1 = APP_NAME + "_rename_1";
+    create_app(app_name_1);
+    auto app = find_app(app_name_1);
```
[the remark this answers is not in the document]
Comment (GehaFearless) [reply src/meta/test/meta_app_operation_test.cpp:816]: `CHECK` and add some message is better.

[the remark this answers is not in the document]
Comment (acelyc111) [reply src/meta/test/meta_app_operation_test.cpp:816]: In gtest CHECK will cause the process crash and no message left, but ASSERT_* will not.

File: src/meta/server_state.cpp [hunk 18]
```diff
@@ -1259,6 +1259,102 @@ void server_state::drop_app(dsn::message_ex *msg)
...
+
+    do_update_app_info(app_path, ainfo, [this, app_id, rpc](error_code ec) mutable {
+        const auto &new_app_name = rpc.request().new_app_name;
+        const auto &old_app_name = rpc.request().old_app_name;
+        zauto_write_lock l(_lock);
+
+        if (ERR_OK != ec) {
+            _exist_apps.erase(new_app_name);
+            CHECK(false,
```
Comment (acelyc111) [inline src/meta/server_state.cpp:1286]: This is will cause the process crash, you can use `CHECK_EQ(ERR_OK, ...)` as before, `_exist_apps.erase(new_app_name);` is meanless because the process will crash soon.

File: src/meta/server_state.cpp [hunk 19]
```diff
@@ -1259,6 +1259,102 @@ void server_state::drop_app(dsn::message_ex *msg)
     }
 }
 
+void server_state::rename_app_unlock(configuration_rename_app_rpc rpc)
+{
+    zauto_write_lock l(_lock);
```
Comment (acelyc111) [inline src/meta/server_state.cpp:1264]: Since the function requires the lock itself, the function name should be `do_rename_app` (or sth like that).

Further more, you can remove the lock if you pass `rpc`, `app_id`, and call `_exist_apps.emplace(...)` in rename_app?

Comment (GehaFearless) [reply src/meta/server_state.cpp:1264]: I see. I think merge two functions into one is better.

File: src/meta/server_state.cpp [hunk 20]
```diff
@@ -1259,6 +1259,87 @@ void server_state::drop_app(dsn::message_ex *msg)
...
+    const auto &new_app_name = rpc.request().new_app_name;
+    LOG_INFO_F(
+        "rename app request, old_app_name({}), new_app_name({})", old_app_name, new_app_name);
+
+    zauto_write_lock l(_lock);
+    auto target_app = get_app(old_app_name);
+
+    if (target_app == nullptr) {
+        response.err = ERR_APP_NOT_EXIST;
+        response.hint_message = fmt::format("ERROR: app({}) not exist. check it!", old_app_name);
```
Comment (acelyc111) [inline src/meta/server_state.cpp:1277]:
```suggestion
        response.hint_message = fmt::format("ERROR: app({}) not exist.", old_app_name);
```

File: src/meta/server_state.cpp [hunk 21]
```diff
@@ -1259,6 +1259,87 @@ void server_state::drop_app(dsn::message_ex *msg)
...
+        return;
+    }
+
+    switch (target_app->status) {
+    case app_status::AS_AVAILABLE: {
+        if (_exist_apps.find(new_app_name) != _exist_apps.end()) {
+            response.err = ERR_INVALID_PARAMETERS;
+            response.hint_message =
+                fmt::format("ERROR: app({}) already exist! check it!", old_app_name);
```
Comment (acelyc111) [inline src/meta/server_state.cpp:1286]:
```suggestion
                fmt::format("ERROR: app({}) already exist!", new_app_name);
```

File: src/meta/server_state.cpp [hunk 22]
```diff
@@ -1259,6 +1259,87 @@ void server_state::drop_app(dsn::message_ex *msg)
...
+    auto ainfo = *(reinterpret_cast<app_info *>(target_app.get()));
+    ainfo.app_name = new_app_name;
+    auto app_path = get_app_path(*target_app);
+
+    target_app->app_name = new_app_name;
+    _exist_apps.emplace(new_app_name, target_app);
+
+    do_update_app_info(app_path, ainfo, [this, app_id, rpc](error_code ec) mutable {
+        const auto &new_app_name = rpc.request().new_app_name;
+        const auto &old_app_name = rpc.request().old_app_name;
```
Comment (acelyc111) [inline src/meta/server_state.cpp:1323]: how about pass them though?

File: src/meta/server_state.cpp [hunk 23]
```diff
@@ -1259,6 +1259,87 @@ void server_state::drop_app(dsn::message_ex *msg)
...
+        zauto_write_lock l(_lock);
+
+        CHECK_EQ_MSG(ec,
+                     ERR_OK,
+                     "update remote app info failed: app_id={}, old_app_name={}, new_app_name={}",
+                     app_id,
+                     old_app_name,
+                     new_app_name);
```
Comment (acelyc111) [inline src/meta/server_state.cpp:1331]:
```suggestion
        CHECK_EQ_MSG(ec,
                     ERR_OK,
                     "update remote app info failed: app_id={}, old_app_name={}, new_app_name={}",
                     app_id,
                     old_app_name,
                     new_app_name);

        zauto_write_lock l(_lock);
```

**Review by acelyc111 — APPROVED** [review]
LGTM

File: src/client/replication_ddl_client.cpp [hunk 24]
```diff
@@ -53,6 +53,15 @@
 namespace dsn {
 namespace replication {
 
+#define VALIDATE_TABLE_NAME(app_name)                                                              \
```
Comment (levy5307) [inline src/client/replication_ddl_client.cpp:56]:
```
inline bool validate_table_name(std::string app_name) {
   ...
}
```
I think it's better to implement it by inline func

**Review by levy5307 — APPROVED** [review]
